### DISCUSSION TITLE
Handling checksum errors or other issues (Python)

### DISCUSSION THREAD

**Discussion by gregtws** [zxing-cpp/zxing-cpp#617] [Q&A] [answered]
I'm working on debugging a QR code that won't scan (Python API).  The image itself is small, 260x260, with a logo in the center.   Its a clean image in terms of black/white as its a generated image and not a picture of a printed QR code. Unfortunately, I'm not sure I can share the sample image at this moment (working on that front).

Changing the various decode parameters does not help (binarizers, rotation, is pure, etc).

After testing against the C++ api and `PRINT_DEBUG` I found that it fails due to a ChecksumError.  If I resize the image just right, or apply a light box/gaussian blur it can extract the QR code contents.  I'm guessing its due to the logo causing too many errors in the non-modified image.

Given all of this would it be reasonable to expose a Python API that can say "QR codes may be present" (valid or not)?   Alternately, exposing a lower level API that lets me get to the root cause error (return error flag or full hints api)?  This would let me avoid having to pre-process all images in our pipeline, and only do extra work on likely candidates.

I'm open to other ideas on pre-processing or deciding when to pre-process as well.

Thanks!

**Comment by axxel** [comment]
It would certainly help if you could share the problematic image. Based on your description, I can imagine that the author of the image did not really know what they were doing and simply pasted something 'too big' over the qrcode symbol for it to be reliably detected. I've seen that in the past. It then can happen that a tiny change may cause the grid sampling to be a little bit shifted and maybe just one module gets detected as 1 vs 0 and suddenly the error correction is enough to decode it.

The fix would then be to either increase the EC level (if not at the highest level already) or make the overlay image smaller.

There is the `DecodeHints::setReturnErrors` flag (see `ZXingReader -error`). So the correct and also easiest approach would be to add a `return_errors` parameter to the `read_barcodes` function and add a suitable property to the result type. If you feel like giving it a go, please do. Otherwise, I can add that in the coming days myself.

**Reply by gregtws** [reply]
Wow thanks for the fast reply.  The image origin is from a phish message so it could be accident or intentional (phone can read the code just fine).  Hard to know.

I've a prototype exposing the error as part of the result.   I'll see if I can put a draft PR out later today.  Just need to fixup the formatting (looks like the file hasn't been formatted with cmake-format yet, so it makes the PR much larger than it really is).

**Reply by axxel** [reply, marked as answer]
Thanks to @gregtws the discussed new `return_errors` API is now implemented (#618).
